src/charm.py

```diff
@@ -5,19 +5,176 @@
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            service_name="webui",
+            ports=[
+                ServicePort(name="urlport-http", port=5000),
+                ServicePort(name="grpc", port=9876),
+            ],
+        )
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle database created event."""
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        if not self._container.exists(path=BASE_CONFIG_PATH):
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        config_file_content = self._render_config_file(
+            database_name=DATABASE_NAME, database_url=event.uris.split(",")[0]
+        )
+        if not self._config_file_is_written() or not self._config_file_content_matches(
+            content=config_file_content
+        ):
+            self._write_config_file(
+                content=config_file_content,
+            )
+        self._on_webui_pebble_ready(event)
+
+    @staticmethod
+    def _render_config_file(database_name: str, database_url: str) -> str:
+        """Renders webui configuration file based on Jinja template.
+
+        Args:
+            database_name: Database Name
+            database_url: Database URL.
+
+        Returns:
+            str: Content of the configuration file.
+        """
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template("webuicfg.conf.j2")
+        return template.render(
+            database_name=database_name,
+            database_url=database_url,
+        )
+
+    def _write_config_file(self, content: str) -> None:
+        """Writes configuration file based on provided content.
+
+        Args:
+            content: Configuration file content
+        """
+        self._container.push(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content)
+        logger.info(f"Pushed {CONFIG_FILE_NAME} config file")
+
+    def _config_file_is_written(self) -> bool:
+        """Returns whether the configuration file is written."""
+        return bool(self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"))
+
+    def _config_file_content_matches(self, content: str) -> bool:
+        """Returns whether the config file content matches the provided content.
+
+        Returns:
+            bool: Whether the config file content matches
+        """
+        existing_content = self._container.pull(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}")
+        return bool(existing_content.read() == content)
+
+    def _on_webui_pebble_ready(
+        self,
+        event: Union[PebbleReadyEvent, DatabaseCreatedEvent],
```

Comment:
I think using `EventBase` instead of listing all the event types is a bit better, because we do not depend on anything more than what is declared in `EventBase`, and future changes to the charm will not need adding new event types to this list.

Reply:
Done